vm/src/builtins/function.rs

```diff
@@ -482,6 +482,28 @@ impl PyBoundMethod {
     fn module(&self, vm: &VirtualMachine) -> Option<PyObjectRef> {
         vm.get_attribute(self.function.clone(), "__module__").ok()
     }
+
+    #[pyproperty(magic)]
+    fn qualname(&self, vm: &VirtualMachine) -> PyResult {
+        if self
+            .function
+            .isinstance(&vm.ctx.types.builtin_function_or_method_type)
+        {
+            // Special case: we work with `__new__`, which is not really a method.
+            // It is a function, so its `__qualname__` is just `__new__`.
+            // We need to add object's part manually.
+            // Note: at the moment, `__new__` in the form of `tp_new_wrapper`
+            // is the only instance of `builtin_function_or_method_type`.
+            let obj_name = vm.get_attribute_opt(self.object.clone(), "__qualname__")?;
+            let obj_name: Option<PyStrRef> = obj_name.and_then(|o| o.downcast().ok());
+            return Ok(vm.ctx.new_str(format!(
+                "{}.__new__",
+                obj_name.as_ref().map_or("?", |s| s.as_str())
```

Comment:
I was not sure about this `?` feature, but I've copied the `repr` implementation.
Maybe we should `match obj_name` and `return Err` on `None`?

Reply:
How CPython handle it? I didn't looked up the history but I guess that happened just to avoid errors at that time due to lack of implementations.

Reply:
In CPython it is just passed to `self.func`: https://github.com/python/cpython/blob/ddf8ae31a0f371eff2db14c7f7a45976b86d56ea/Objects/classobject.c#L201

`repr` is also quite different: https://github.com/python/cpython/blob/ddf8ae31a0f371eff2db14c7f7a45976b86d56ea/Objects/classobject.c#L274-L300 It also has no `?`

Reply:
I can see `?` here:
https://github.com/python/cpython/blob/ddf8ae31a0f371eff2db14c7f7a45976b86d56ea/Objects/classobject.c#L280

Reply:
Yes, indeed! I've missed that. 

Anyway, let's leave `repr` out of scope (as it is for this task). 
What should we do with `__qualname__`? 🙂 

I see several things:
1. Leave it as is, until someone report a case where this code is wrong
2. Switch `map_or` to `unwrap()`, because all object must have this property, this can cause bugs if some objects accidentally don't have it
3. ???

Reply:
I think this branch itself is not the CPython way anyway. So any approach looks acceptable. When we handle `__new__` correctly, the whole branch will be gone. isn't it?

Reply:
Yes, it only executes for `__new__` on several builtin types, others are just proxied to `vm.get_attribute(self.function.clone(), "__qualname__")` like CPython does. I can add a `TODO` for future to fix this special case on a lower-level.